core/documents-manager.php

```diff
@@ -0,0 +1,214 @@
+	}
+
+	/**
+	 * @param int $post_id
+	 *
+	 * @return Document
+	 */
+	public function get( $post_id ) {
+		if ( ! isset( $this->documents[ $post_id ] ) ) {
+			$doc_type = get_post_meta( $post_id, Document::TYPE_META_KEY, true );
+
+			$doc_type_class = $this->get_document_type( $doc_type );
+			$this->documents[ $post_id ] = new $doc_type_class( [
+				'post_id' => $post_id,
+			] );
+		}
+
+		return $this->documents[ $post_id ];
+	}
+
+	public function get_document_type( $type ) {
+		return isset( $this->types[ $type ] ) ? $this->types[ $type ] : $this->types['post'];
+	}
+
+	/**
+	 * @param string $class_name
+	 * @param array $post_data
+	 * @param array $meta_data
+	 *
+	 * @return Document
+	 */
+	public function create( $class_name, $post_data = [], $meta_data = [] ) {
+		if ( empty( $post_data['post_title'] ) ) {
+			$post_data['post_title'] = __( 'Elementor', '' );
+			$update_title = true;
+		}
+
+		$post_id = wp_insert_post( $post_data );
+
+		if ( ! empty( $update_title ) ) {
+			$post_data['ID'] = $post_id;
+			$post_data['post_title'] .= ' #' . $post_id;
+			wp_update_post( $post_data );
+		}
+
+		add_post_meta( $post_id, '_elementor_edit_mode', 'builder' );
+
+		foreach ( $meta_data as $key => $value ) {
+			add_post_meta( $post_id, $key, $value );
+		}
+
+		/** @var Document $document */
+		$document = new $class_name( [
+			'post_id' => $post_id,
+		] );
+
+		$document->save_type();
+
+		return $document;
+	}
+
+	/**
+	 * @since 1.0.0
+	 * @access public
+	 */
+	public function	ajax_save( $request ) {
+		if ( empty( $request['post_id'] ) ) {
+			throw new \Exception( 'no_post_id' );
+		}
+
+		$document = $this->get( $request['post_id'] );
+
+		if ( ! $document->is_built_with_elementor() || ! $document->is_editable_by_current_user() ) {
+			wp_send_json_error( new \WP_Error( 'no_access' ) );
+		}
+
+		$status = DB::STATUS_DRAFT;
+
+		if ( isset( $request['status'] ) && in_array( $request['status'], [ DB::STATUS_PUBLISH, DB::STATUS_PRIVATE, DB::STATUS_PENDING, DB::STATUS_AUTOSAVE ] , true ) ) {
+			$status = $request['status'];
+		}
+
+		if ( DB::STATUS_AUTOSAVE === $status ) {
+			if ( ! defined( 'DOING_AUTOSAVE' ) ) {
+				define( 'DOING_AUTOSAVE', true );
+			}
+
+			// If the post is a draft - save the `autosave` to the original draft.
+			// Allow a revision only if the original post is already published.
+			if ( in_array( $document->get_post()->post_status, [ DB::STATUS_PUBLISH, DB::STATUS_PRIVATE ], true ) ) {
+				$document = $document->get_autosave( 0, true );
+			}
+		}
+
+		$data = [
+			'elements' => json_decode( stripslashes( $request['data'] ), true ),
+		];
+
+		$data['settings']['post_status'] = $status;
+
+		$document->save( $data );
+
+		$return_data = [
+			'config' => [
+				'last_edited' => $document->get_last_edited(),
+				'wp_preview' => [
+					'url' => $document->get_wp_preview_url(),
+				],
+
+			],
+		];
+
+		/**
+		 * Filters the ajax data returned when saving the post on the builder.
```

Comment:
Add filter summary above the description.

**Summary:** Returned documents ajax saved data.